packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,558 @@
+        if (send.fun != core::Names::export_()) {
+            return tree;
+        }
+
+        if (send.numPosArgs() != 1) {
+            // an error will have been raised in the packager pass
+            return tree;
+        }
+
+        auto *lit = ast::cast_tree<ast::ConstantLit>(send.getPosArg(0));
+        if (lit == nullptr || lit->symbol == core::Symbols::StubModule()) {
+            // We don't raise an explicit error here, as this is one of two cases:
+            //   1. Export is given a non-constant argument
+            //   2. The argument failed to resolve
+            // In both cases, errors will be raised by previous passes.
+            return tree;
+        }
+
+        auto &db = ctx.state.packageDB();
+        if (lit->symbol.isClassOrModule()) {
+            auto sym = lit->symbol.asClassOrModuleRef();
+            checkExportPackage(ctx, send.loc, db.getPackageNameForFile(sym.data(ctx)->loc().file()));
+            recursiveExportSymbol(ctx, true, sym);
+
+            exportParentNamespace(ctx, sym.data(ctx)->owner);
+        } else if (lit->symbol.isFieldOrStaticField()) {
+            auto sym = lit->symbol.asFieldRef();
+            checkExportPackage(ctx, send.loc, db.getPackageNameForFile(sym.data(ctx)->loc().file()));
+            sym.data(ctx)->flags.isExported = true;
+
+            // When exporting a field, we also export its parent namespace. This is a bit of a hack, and it would be
+            // great to remove this, but this was the behavior of the previous packager implementation.
+            // TODO: can we get away with only marking the owner as exported, and not everything up to the package?
+            exportParentNamespace(ctx, sym.data(ctx)->owner);
+
+            if (sym.data(ctx)->flags.isStaticFieldTypeAlias) {
+                if (auto e = ctx.beginError(send.loc, core::errors::Packager::ExportingTypeAlias)) {
+                    e.setHeader("Exporting a type alias is not allowed in package specifications");
+                    e.addErrorLine(sym.data(ctx)->loc(), "Originally defined here");
+                }
+            }
+        }
+
+        return tree;
+    }
+
+    ast::ExpressionPtr preTransformClassDef(core::MutableContext ctx, ast::ExpressionPtr tree) {
+        auto &original = ast::cast_tree_nonnull<ast::ClassDef>(tree);
+
+        if (original.symbol == core::Symbols::root()) {
+            return tree;
+        }
+
+        setPackageLocs(ctx, original.name.loc(), original.symbol);
+
+        return tree;
+    }
+
+    static ast::ParsedFile run(core::GlobalState &gs, ast::ParsedFile f) {
+        if (!f.file.data(gs).isPackage()) {
+            return f;
+        }
+
+        auto pkgName = gs.packageDB().getPackageNameForFile(f.file);
+        if (!pkgName.exists()) {
+            return f;
+        }
+
+        const auto &package = gs.packageDB().getPackageInfo(pkgName);
+        if (!package.exists()) {
+            return f;
+        }
```

Comment:
when would `f` be a package file, but either `pkgName` or `package` doesn't exist?

Reply:
I'm honestly not sure how this could happen, but the types allow for it so I was being extra careful. I'd be happy to replace the second conditional with an `ENFORCE`.

Reply:
Either way. I’m happy to also add the ENFORCE but also keep the return